feat(forms): introduce InputGroup component

austin94 opened the pull request:
## Description

This PR introduces the `InputGroup` component. It allows Buttons (and other elements) to be prepended, appended around an `Input` element.

<img width="543" alt="Screen Shot 2020-08-18 at 11 19 55 AM" src="https://user-images.githubusercontent.com/4030377/90551447-33214480-e146-11ea-9281-21bcab83606f.png">

```tsx
<Field>
  <Label>Input group</Label>
  <InputGroup>
    <Button>A</Button>
    <Input />
    <Button>B</Button>
  </InputGroup>
</Field>
```

### `focusInset` styling

To ensure consistent focus styling I'm using a context object to inform the `Input` element to default to use `focusInset` styling. Since the `Prepend/Append` elements can receive any element it is up to consumers to apply the props to the element if they're using a Garden `Button`.

### Field sizing

The `InputGroup` element disrupts the current styling between labels, hints, messages, and input. To ensure consistent styling I've included an `isCompact` prop to `InputGroup` which is also passed to the `Input` through context.

## Checklist

- [x] :ok_hand: design updates are Garden Designer approved (add the
      designer as a reviewer)
- [x] :globe_with_meridians: Styleguidist demo is up-to-date (`yarn start`)
- [x] :arrow_left: renders as expected with reversed (RTL) direction
- [x] :metal: renders as expected with Bedrock CSS (`?bedrock`)
- [x] :wheelchair: analyzed via [axe](https://www.deque.com/axe/) and evaluated using VoiceOver
- [x] :guardsman: includes new unit tests
- [x] :memo: tested in Chrome, Firefox, Safari, Edge, and IE11

m-lai reviewed (approved):
Looks good to me. I think if we allow multiple buttons/elements then `Prepend` and `Append` make sense. Sketch symbol naming will still follow the Split Button naming format I reckon, so it's easier for designers to use:

1. input-group-default/1. primary/1. default/1. default
1. input-group-default/1. primary/1. default/2a. hover-input
1. input-group-default/1. primary/1. default/2b. hover-button
1. input-group-default/1. primary/1. default/3. active-button
1. input-group-default/1. primary/1. default/4a. focus-input

austin94 commented:
@jzempel weird, this is only showing on iOS mobile.

<img width="282" alt="Screen Shot 2020-08-20 at 1 42 14 PM" src="https://user-images.githubusercontent.com/4030377/90823961-841e6d80-e2eb-11ea-9592-ed65d12135f0.png">

Looking better to me know.

austin94 commented on packages/forms/src/styled/input-group/StyledInputGroup.ts:
```diff
@@ -0,0 +1,80 @@
+}
+
+const positionStyles = (props: ThemeProps<DefaultTheme> & IStyledInputGroupProps) => {
+  const topMargin = math(`${props.theme.space.base} * ${props.isCompact ? '1px' : '2px'}`);
```
Updated to remove the `math` utility.

austin94 commented on packages/forms/src/styled/input-group/StyledInputGroup.ts:
```diff
@@ -0,0 +1,80 @@
+  'data-garden-id': COMPONENT_ID,
+  'data-garden-version': PACKAGE_VERSION
+})<IStyledInputGroupProps>`
+  display: flex;
```
> Is there an advantage for making this occupy the entire block?

Our inputs are `display: block` by default I thought. Would we want these groups to be inline?

> should direction be getting applied based on RTL?

The parent `Field` already applies that for us!

austin94 commented on packages/forms/src/styled/input-group/StyledInputGroup.ts:
```diff
@@ -0,0 +1,80 @@
+      flex: 1 1 auto;
+      margin-top: 0;
+      margin-bottom: 0;
+      width: 1px;
```
This overrides the default `width: 100%` of the `StyledTextInput`

<img width="919" alt="Screen Shot 2020-08-24 at 2 37 49 PM" src="https://user-images.githubusercontent.com/4030377/91099052-76ccf000-e617-11ea-8fb3-48c5dd2eabbf.png">

I've added an inline comment for this.

austin94 commented on packages/forms/examples/input-group.md:
```diff
@@ -0,0 +1,28 @@
+        Copy
+      </Button>
+    </InputGroup.Append>
+  </InputGroup>
```
This is handled correctly I believe. The styling is applied when a sibling of the `StyledLabel`

```
${StyledLabel} + & {
  /** margin-top styles */
}
```

austin94 replied:
I've added a validation message to the new knobs example to show this.

austin94 commented on packages/forms/examples/input-group.md:
```diff
@@ -0,0 +1,28 @@
+
+<Field>
+  <Label>Input Group</Label>
+  <InputGroup>
```
I've added a knobs section with an `isCompact` toggle

austin94 commented on packages/forms/src/styled/input-group/StyledPrepend.ts:
```diff
@@ -0,0 +1,48 @@
+  'data-garden-version': PACKAGE_VERSION
+})`
+  display: flex;
+  z-index: 2;
```
This was tricky to bring over, but I think I got it working. We need to start out the `z-index` at `1` with the `InputGroup` to beat the `StyledTextInput` styling.

I've added similar z-index logic to the `Prepend` and `Append` components.

![image](https://user-images.githubusercontent.com/4030377/91100305-ca403d80-e619-11ea-8ee7-fe109748e843.png)

jzempel commented on packages/forms/src/styled/input-group/StyledAppend.ts:
```diff
@@ -0,0 +1,57 @@
+    z-index: 2;
+  }
+
+  &:disabled {
+    z-index: -1;
+  }
```
Disabled styling doesn't appear to be playing as expected. This div cannot be disabled. But when it's child button is disabled, the hover styling gets wonky. Styling should reflect, to some degree, the disabled styling seen with https://zendeskgarden.github.io/react-components/buttons/#!/Basic/9.

jzempel replied:
Could be that the example should be utilizing a `ButtonGroup` within the `InputGroup` in order to get the styling right?

austin94 replied:
> disabled styling doesn't appear to be playing as expected

I was accidentally applying `z-index` to the wrapping div. This now targets the child elements correctly.

austin94 replied:
I'm not sure using a `ButtonGroup` would work here. It includes semantic roles and keyboard interaction that diverges from this pattern.

jzempel commented on packages/forms/src/styled/input-group/StyledInputGroup.ts:
```diff
@@ -0,0 +1,82 @@
+      flex: 1 1 auto;
+      margin-top: 0;
+      margin-bottom: 0;
+      width: 1px; /* [1] */
```
I think you might want `width: auto;` here to override and reset.

jzempel commented on packages/forms/src/styled/input-group/StyledAppend.ts:
```diff
@@ -0,0 +1,57 @@
+
+  &:hover,
+  &:active {
+    z-index: 2;
```
> This was tricky to bring over, but I think I got it working. We need to start out the z-index at 1 with the InputGroup to beat the StyledTextInput styling.

I'm not seeing that `StyledTextInput` has a `z-index`.

austin94 replied:
Without starting at `z-index: 1` we have overlay issues with the input 

![image](https://user-images.githubusercontent.com/4030377/91207310-6b340480-e6bd-11ea-8f8d-0905209a6a8f.png)

jzempel replied:
Got it. In that case, wouldn't `StyledTextInput` start at `z-index: -1` and then apply `z-index: 1` on hover/active?

austin94 replied:
`z-index: -1` begins to place elements behind the body at that point

![image](https://user-images.githubusercontent.com/4030377/91234972-52d9df00-e6e9-11ea-912c-cc87349b61bf.png)

jzempel replied:
This is why resetting the stacking context by applying `z-index: 0` to the `InputGroup` wrapper is so important. Once that is done, the children only need to receive either `z-index: 1` or `z-index: -1`.

jzempel replied:
See https://developer.mozilla.org/en-US/docs/Web/CSS/z-index#Values for details.

jzempel commented on packages/forms/src/styled/input-group/StyledInputGroup.ts:
```diff
@@ -0,0 +1,82 @@
+  'data-garden-id': COMPONENT_ID,
+  'data-garden-version': PACKAGE_VERSION
+})<IStyledInputGroupProps>`
+  display: inline-flex;
```
> Our inputs are display: block by default I thought.

Garden's text inputs are `display: inline-block`, making `inline-flex` the analog.

austin94 commented:
@jzempel I've updated this to remove the `Prepend/Append` components as requested and added a `disabled` knob to the example.

I'm still experiencing some z-index weirdness with the text input due to the negative margin we apply. I've bumped the z-index to fix this, but want to double check with you.

jzempel commented on packages/forms/examples/input-group.md:
```diff
@@ -0,0 +1,83 @@
+            <Field>
+              <Toggle
+                checked={state.isDisabled}
+                onChange={event => setState({ isDisabled: event.target.checked })}
```
The disabled state should also apply to the input field.

austin94 replied:
Updated.

jzempel commented on packages/forms/src/styled/input-group/StyledInputGroup.ts:
```diff
@@ -0,0 +1,123 @@
+
+    & > button:hover,
+    & > button:active {
+      z-index: 2;
```
Once stacking context is applied to the `InputGroup`, this should be z-index: 1. Also the rule here should be for all of:
```css
& > ${StyledInput}:hover,
& > button:hover,
& > ${StyledInput}:active,
& > button:active {
  z-index: 1;
}
```

austin94 replied:
Stacking context has been updated.

jzempel commented on packages/forms/src/styled/input-group/StyledInputGroup.ts:
```diff
@@ -0,0 +1,123 @@
+      margin-${props.theme.rtl ? 'right' : 'left'}: -${props.theme.borderWidths.sm};
+    }
+
+    & > button {
+      z-index: 1;
+    }
```
Once stacking context is applied to the `InputGroup`, replace this with:

```css
& > ${StyledInput} {
  z-index: -1;
}
```

austin94 replied:
Stacking context has been updated.

jzempel commented on packages/forms/src/styled/input-group/StyledInputGroup.ts:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    & > button:disabled {
+      z-index: 0;
```
Once stacking context is applied to the `InputGroup`, this should be `z-index: -1`.

austin94 replied:
Stacking context has been updated.

jzempel commented on packages/forms/src/styled/input-group/StyledInputGroup.ts:
```diff
@@ -0,0 +1,123 @@
+      border-top-width: 0;
+      border-bottom-width: 0;
+    }
+
```
Once stacking context is applied to `InputGroup`, add:

```css
& > ${StyledInput} {
  z-index: -2;
}
```

This is the one case that drops outside the [-1 : 1] threshold.

austin94 replied:
Stacking context has been updated.

jzempel commented on packages/forms/src/styled/input-group/StyledInputGroup.ts:
```diff
@@ -0,0 +1,123 @@
+      border-bottom-width: 0;
+    }
+
+    & > button:first-of-type:not(:last-of-type) {
+      margin-${props.theme.rtl ? 'right' : 'left'}: 0;
+      border-top-${props.theme.rtl ? 'left' : 'right'}-radius: 0;
+      border-bottom-${props.theme.rtl ? 'left' : 'right'}-radius: 0;
+    }
+
+    & > button:last-of-type:not(:first-of-type) {
+      border-top-${props.theme.rtl ? 'right' : 'left'}-radius: 0;
+      border-bottom-${props.theme.rtl ? 'right' : 'left'}-radius: 0;
+    }
+
+    & > button:not(:first-of-type):not(:last-of-type) {
+      border-radius: 0;
+    }
```
These should not be qualified with `button`. Removing elements breaks styling.

austin94 replied:
I've updated this to use `:first-child` and `:last-child` pseudo selectors.

jzempel commented:
blerg, these false-positive deploy tests should get solved by the `@zendeskgarden/scripts` update in #845. In the meantime, I'll pull and test locally.

austin94 commented on packages/forms/examples/input-group.md:
```diff
@@ -0,0 +1,83 @@
+- `InputGroup` Accepts all `div` attributes and the following props:
+  - [`isCompact`] Apply compact styling
+- `InputGroup.Prepend` Accepts all `div` attributes
+- `InputGroup.Append` Accepts all `div` attributes
```
Updated both.

jzempel replied:
[nit] the PR description is still referring to Append/Prepend.

austin94 replied:
Missed the description shown in the image background. Image has been updated now.
